fix: Transform network failures into specific TLS timeout

thiyaguk09 opened the pull request:
## Description

> This change refines error handling in `util.makeRequest` to intercept and transform common low-level network failures (`ECONNRESET`, `ETIMEDOUT`, `"timed out"`, `"TLS handshake"`) into **three distinct and specific standard `Error` messages**.
>
>The following raw errors are now intercepted:
>* `ECONNRESET`
>* `ETIMEDOUT`
>* Generic messages containing `"timed out"`
>* Generic messages containing `"TLS handshake"`
>
>Instead of a single message, developers now receive tailored diagnostic information: a specific message for **TLS/CPU starvation issues**, one for **network timeouts**, and one for **connection resets**. This provides clearer guidance for debugging. The original stack trace is preserved on the new `Error` object. This is an additive change with no breaking impacts, validated by a consolidated, data-driven unit test suite.
>
>By transforming these errors and preserving the original stack trace, we prevent the propagation of cryptic, low-level network codes and provide developers with a clear, unified diagnostic message tailored to the type of connection failure experienced.
>

## Impact

>The impact of this change is primarily positive, improving the developer experience:
>
>* **Improved Error Diagnostics (Granular):** Developers receive **three distinct, specific messages** instead of a single ambiguous one. For instance, a `TLS handshake` error receives a special message related to **CPU starvation** to guide performance debugging.
>* **Consistent Error Handling:** Facilitates easier integration with custom error retry and logging mechanisms by providing a predictable, standard `Error` structure (augmented with the original stack trace) rather than a raw, non-standard network error object.
>* **No Breaking Changes:** This is a purely additive fix that catches and transforms errors that would have been thrown anyway. It does not alter the successful path for requests.
>

## Testing

>**Yes, unit tests were added.**
>
>* A dedicated unit test suite, **`Network Connectivity Errors`**, was created under `makeAuthenticatedRequestFactory` to validate the new transformation logic.
>* The test structure was consolidated into a single, **data-driven `forEach` loop** that verifies the mapping of the four common failure modes to the **three distinct output messages** defined in the `UtilExceptionMessages` enum.
>
>* A single test loop replaced four separate tests, covering all conditions:
>    * `should transform raw ECONNRESET into specific network error`
>    * `should transform raw "TLS handshake" into specific network error`
>    * `should transform raw generic "timed out" into specific network error`
>    * `should transform raw ETIMEDOUT into specific network error`
>
>**Tests Changed?** No existing tests were modified.
>
>**Breaking Changes?** No breaking changes are necessary.
>

## Additional Information
>
>*  **Error Object Change:** The transformation logic was simplified to augment a **standard JavaScript `Error`** object.
>* **Test Structure:** The tests were consolidated into a single `forEach` loop for improved clarity and maintainability.
>* **Stubbing:** The `authClient` was stubbed to guarantee successful authorization, forcing execution into the network path where the error injection and transformation occur, preventing test timeouts.
> The logic ensures that if an error is transformed, the original stack trace (`err.stack`) is preserved on the new `Error` object, allowing developers to debug the source of the failure effectively.

## Checklist

- [ ] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/nodejs-storage/issues/new/choose) before writing your code! That way we can discuss the change, evaluate designs, and agree on the general idea
- [ ] Ensure the tests and linter pass
- [ ] Code coverage does not decrease
- [ ] Appropriate docs were updated
- [ ] Appropriate comments were added, particularly in complex areas or places that require background
- [ ] No new warnings or issues will be generated from this change

Fixes #

thiyaguk09 committed:
Fix: Transform network connectivity errors into TLS timeout ApiError

Transforms raw network errors (ECONNRESET, ETIMEDOUT, timed out, and TLS
handshake) into a specific ApiError (code 408) with a descriptive
message regarding potential CPU starvation.

This prevents misleading error propagation from the underlying request
library.

thiyaguk09 committed:
fix

thiyaguk09 committed:
fix: Differentiate retry codes for network errors

Splits network error handling: uses 408 for timeouts (timed out,
ETIMEDOUT, TLS handshake) and 503 for connection resets (ECONNRESET) to
improve retry logic accuracy.

v-pratap commented on src/nodejs-common/util.ts:
```diff
@@ -912,6 +912,38 @@ export class Util {
+          if (
+            err &&
+            (err.message?.toLowerCase().includes('tls handshake') ||
+              err.message?.toLowerCase().includes('timed out') ||
```
The err.message.toLowerCase() call is repeated multiple times inside this block. It would be slightly more efficient to call it once, store the result in a variable, and then use that variable in the if conditions.

v-pratap commented on test/nodejs-common/util.ts:
```diff
@@ -1189,6 +1189,100 @@ describe('common/util', () => {
+          authClient.authorizeRequest = async () => {
+            throw error;
+          };
+        });
```
The new tests are very similar in code, with a lot of duplicated code. Is it possible to reuse the code?

thiyaguk09 replied:
Thank you for your input. I will review it and respond to you later.

thiyaguk09 commented on src/nodejs-common/util.ts:
```diff
@@ -912,6 +912,38 @@ export class Util {
+              // TLS Handshake, ETIMEDOUT, or generic 'timed out'
+              // Mapped to 408 (Request Timeout) as a signal for retrying
+              code = 408;
+              message =
```
Thank you for your input. I will review it and respond to you later.

thiyaguk09 commented on src/nodejs-common/util.ts:
```diff
@@ -912,6 +912,38 @@ export class Util {
+            let message: string;
+            if (err.message.toLowerCase().includes('econnreset')) {
+              // ECONNRESET (Connection reset by peer) implies temporary service unavailability
+              code = 503;
```
That's an insightful point! You're right, TCP/socket errors like ECONNRESET are fundamental network issues and shouldn't be incorrectly treated as HTTP status codes. They reflect a connection failure, not necessarily an application-level server error.

I will check on it and come back.

thiyaguk09 committed:
fix: Transform network errors into specific TLS/CPU starvation Error

Converts raw ECONNRESET, ETIMEDOUT, and TLS handshake failures into a
standard Error object with an informative message. This helps diagnose
CPU starvation or misleading 401 errors.

thiyaguk09 committed:
refactor: Combine duplicate TLS error tests into data-driven suite

Replaces repetitive test cases in `makeAuthenticatedRequest` and
`makeRequest` with a single, data-driven test loop. This verifies all
conditions (ECONNRESET, ETIMEDOUT, "timed out", "TLS handshake") with
reduced code duplication and improved maintenance.
```

thiyaguk09 committed:
Code refactor

ddelgrosso1 commented:
I'm not really sure why we are forcing things such as `ECONNRESET`, `ETIMEDOUT` into a TLS error. They may or may not be related to TLS. I think this gives a false impression to the end user. I think we need to rethink what it is we are trying to accomplish here.

thiyaguk09 commented:
> I'm not really sure why we are forcing things such as `ECONNRESET`, `ETIMEDOUT` into a TLS error. They may or may not be related to TLS. I think this gives a false impression to the end user. I think we need to rethink what it is we are trying to accomplish here.

I agree completely—the current grouping is misleading. Thanks for the feedback! 

I've updated the error handling to **distinguish these failures**:
* `"tls handshake"` for specific **protocol issues**.
* `"etimedout"` / `"timed out"` for **network timeouts**.
* `"econnreset"` for **connection resets**.

This provides a much more accurate diagnosis for the end-user.

thiyaguk09 committed:
Fix: Improve error handling to distinguish network errors from TLS failures

Separates specific network transport errors (`ETIMEDOUT`, `ECONNRESET`)
from genuine TLS handshake failures.

The previous approach incorrectly categorized lower-level connection
issues as "TLS errors," leading to misleading diagnostics for end-users.
This change ensures accurate reporting based on the error pattern:
- "tls handshake": Protocol/certificate issue.
- "etimedout" / "timed out": Network timeout/availability.
- "econnreset": Connection forcefully reset by host/intermediary.
